### PR TITLE
Configured parse server and implemented adding tasks

### PR DESCRIPTION
**Pull request by sosaagbontaen**
I was able to connect the Task Class, with basic properties, to a Parse server. I also implemented a basic UI that allows users to add a new task their to-do list.

### CODE REVIEW

File: CentralityiOS/CentralityiOS/AddTaskModalViewController.m [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        [self alert:@"Invalid Task Name" messageLabel:@"Task Name field is empty" label:@"OK"];
+        return;
+    }
+    
+    [self dismissViewControllerAnimated:YES completion:^{}];
+    
+    TaskObject *newTask = [TaskObject new];
+    
+    //newTask.taskID = User.taskCounter;
+    //newTask.dueDate;
```
Comment (mayguan) [inline CentralityiOS/CentralityiOS/AddTaskModalViewController.m]: Please try not to commit commented out code!

File: CentralityiOS/CentralityiOS/AddTaskModalViewController.m [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    newTask.taskTitle = self.taskTitleInput.text;
+    newTask.taskDesc = self.taskDescInput.text;
+    newTask.isCompleted = false;
+    //newTask.category;
```
Comment (mayguan) [inline CentralityiOS/CentralityiOS/AddTaskModalViewController.m]: Same here.

File: CentralityiOS/CentralityiOS/AddTaskModalViewController.m [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+/*
+#pragma mark - Navigation
+
+// In a storyboard-based application, you will often want to do a little preparation before navigation
+- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
+    // Get the new view controller using [segue destinationViewController].
+    // Pass the selected object to the new view controller.
+}
+*/
```
Comment (mayguan) [inline CentralityiOS/CentralityiOS/AddTaskModalViewController.m]: Same here.

File: CentralityiOS/CentralityiOS/AppDelegate.m [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+        NSString *path = [[NSBundle mainBundle] pathForResource: @"Keys" ofType: @"plist"];
+        NSDictionary *dict = [NSDictionary dictionaryWithContentsOfFile: path];
+
+        NSString *applicationId = [dict objectForKey: @"applicationId"];
+        NSString *clientKey = [dict objectForKey: @"clientKey"];
+        
+        configuration.applicationId = [NSString stringWithFormat:@"%@", applicationId];
+        configuration.clientKey = [NSString stringWithFormat: @"%@", clientKey];
+        configuration.server = @"https://parseapi.back4app.com";
```
Comment (mayguan) [inline CentralityiOS/CentralityiOS/AppDelegate.m]: Consider using global variables for the keys and url here.

File: CentralityiOS/CentralityiOS/AppDelegate.m [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+
+- (void)application:(UIApplication *)application didDiscardSceneSessions:(NSSet<UISceneSession *> *)sceneSessions {
+    // Called when the user discards a scene session.
+    // If any sessions were discarded while the application was not running, this will be called shortly after application:didFinishLaunchingWithOptions.
+    // Use this method to release any resources that were specific to the discarded scenes, as they will not return.
+}
+
+
```
Comment (mayguan) [inline CentralityiOS/CentralityiOS/AppDelegate.m]: Looks like this is an empty method? Please remove if this is not needed here!

**Review by mayguan — COMMENTED** [review]
A few reminders:
- Try not to commit commented out code or empty methods.
- Consider using global variables for static strings and values.

I've commented at a few places in the code as examples of the above (not exhaustive!). Please address before committing!

Also, I know the first commit is usually bigger but consider breaking future pull requests/commits into smaller ones for easier review and faster iteration!

**Comment by mayguan** [comment]
Could you please add a screenshot or screen recording for the basic UI you built in this commit?

File: .gitignore [hunk 6]
```diff
@@ -0,0 +1,3 @@
+
```
Comment (isaacwycoff) [inline .gitignore:1]: I think you need to have some more things in your .gitignore, or you'll have a bunch of unnecessary files! `.DS_Store`, for one.

Comment (isaacwycoff) [reply .gitignore:1]: `.xcworkspacedata` is another one that might be unnecessary - there are a couple of ways to go: https://stackoverflow.com/questions/22345575/should-xcworkspace-be-in-gitignore

File: CentralityiOS/CentralityiOS/AppDelegate.m (hunk 5 above)
Comment (isaacwycoff) [reply CentralityiOS/CentralityiOS/AppDelegate.m]: This is a part of the `UIApplicationDelegate` protocol, so I'm pretty sure it needs to be here.

File: CentralityiOS/CentralityiOS/TaskObject.h [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+NS_ASSUME_NONNULL_BEGIN
+@interface TaskObject : PFObject<PFSubclassing>
+
+@property (nonatomic) NSInteger taskID;
+@property (nonatomic, strong) NSDate *dueDate;
+@property (nonatomic, strong) NSString *taskTitle;
+@property (nonatomic, strong) NSString *taskDesc;
+@property (nonatomic) BOOL isCompleted;
+@property (nonatomic, strong) NSString *category;
+//@property (nonatomic, strong) User *owner;
```
Comment (isaacwycoff) [inline CentralityiOS/CentralityiOS/TaskObject.h]: if you don't need it, get rid of it for now!

File: CentralityiOS/CentralityiOS/ToDoFeedViewController.m [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+@end
+
+@implementation ToDoFeedViewController
+
+- (void)fetchData{
+    // construct query
+    PFQuery *query = [PFQuery queryWithClassName:@"TaskObject"];
+    [query orderByDescending:@"createdAt"];
+    //[query whereKey:@"likesCount" greaterThan:@100];
```
Comment (isaacwycoff) [inline CentralityiOS/CentralityiOS/ToDoFeedViewController.m]: more commented out code to be deleted.

File: CentralityiOS/CentralityiOS/ToDoFeedViewController.m [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    [self fetchData];
+    
+    self.refreshControl = [[UIRefreshControl alloc] init];
+    [self.refreshControl addTarget:self action:@selector(fetchData) forControlEvents:UIControlEventValueChanged];
+    [self.taskTableView insertSubview:self.refreshControl atIndex:0];
+    
+    // Do any additional setup after loading the view.
+}
+
+/*
```
Comment (isaacwycoff) [inline CentralityiOS/CentralityiOS/ToDoFeedViewController.m]: more commented out code to axe!

if some of this stuff is in-progress, you can break it into separate diffs or create separate branches until it's ready.

**Comment by isaacwycoff** [comment]
I second all of May's comments. The other big one is your `.gitignore` - it's easiest to clean this up at the beginning of the project, so you don't end up with a weird history of binary files getting added or removed. For any file that you didn't explicitly create yourself (like a `.h`, `.m` or `.json` file), if you're not sure if it should be committed or not, look it up. You might also be able to find a good template for `xcode + objective-c` gitignore files that you can use. In general the guideline is: if it's a build artifact or a file that Xcode will regenerate on build, then you shouldn't commit it. (There are some exceptions, but they're pretty rare.)

Pods should _generally_ not be committed. The exception would be if you are modifying a pod inside your project.

**Comment by isaacwycoff** [comment]
It might also not be a bad idea to try your hand at splitting this 1 commit up! I think there are a few ways to split it, and it will make for a nicer and easier to understand history. If you haven't done this in `git` before, it's a good thing to know how to do: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_splitting_a_commit

You've got a fair amount in here that's the basic scaffolding of the app, which makes sense. If I were to split it up, I'd split the basic scaffolding, followed by small commits with any of the files you explicitly added. After that, commits with any initial changes you've made to the files that were generated with templates.

File: CentralityiOS/CentralityiOS/ToDoFeedViewController.m [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    [query orderByDescending:@"createdAt"];
+    //[query whereKey:@"likesCount" greaterThan:@100];
+    query.limit = 20;
+
+    // fetch data asynchronously
+    [query findObjectsInBackgroundWithBlock:^(NSArray *tasks, NSError *error) {
+        if (tasks != nil) {
+            self.arrayOfTasks = tasks;
+        } else {
+            NSLog(@"%@", error.localizedDescription);
```
Comment (isaacwycoff) [inline CentralityiOS/CentralityiOS/ToDoFeedViewController.m:28]: should we still call `reloadData` in this case? i'm not sure.

Comment (sosaagbontaen) [reply CentralityiOS/CentralityiOS/ToDoFeedViewController.m:28]: Removing it here results in an empty table view

**Comment by sosaagbontaen** [comment]
Screenshots and screen recordings of current UI can be found here : https://www.notion.so/centralityios/Configure-Parse-Server-0c46a806e9524a9ea8c4a449b55f71f9

File: CentralityiOS/CentralityiOS/AppDelegate.m [hunk 11]
```diff
@@ -7,20 +7,21 @@
 
 #import "AppDelegate.h"
 #import "Parse/Parse.h"
-
-@interface AppDelegate ()
-@end
+NSString *path;
+NSDictionary *dict;
+NSString *applicationId;
+NSString *clientKey;
```
Comment (mayguan) [inline CentralityiOS/CentralityiOS/AppDelegate.m]: Create global variables for static values. For example,
`static const NSString *kApplicationIdKey = @"applicationId";`
and then use it like
`NSString *applicationId = [dict objectForKey: kApplicationIdKey];`

File: CentralityiOS/CentralityiOS/AppDelegate.m [hunk 12]
```diff
@@ -7,25 +7,28 @@
...
-NSString *path;
-NSDictionary *dict;
-NSString *applicationId;
-NSString *clientKey;
+
+static const NSString *applicationIdObj = @"applicationId";
+static const NSString *clientKeyObj = @"clientKey";
+static NSString *serverURL = @"https://parseapi.back4app.com";
+static NSString *keysFileName = @"Keys";
+static NSString *keysFileExtension = @"plist";
```
Comment (mayguan) [inline CentralityiOS/CentralityiOS/AppDelegate.m:15]: Don't forget to add `const` and prefix variable names with k!

**Review by mayguan — APPROVED** [review]

**Pull request merged**